### PR TITLE
osd/scrub: don't block high-priority scrubs on local resources

### PR DESCRIPTION
**Pull request by ronen-fr**
A scrub can only be initiated if the number of active scrubs performed on the OSD when acting as a primary is below a certain configured limit. After this change - high priority scrubs are not blocked by this limit (although they are still counted towards it).

### CODE REVIEW

**Comment by ronen-fr** [comment]
This is an updated version of https://github.com/ceph/ceph/pull/54348. In this new version, the
scrub initiation sequence is modified to allow high priority scrubs to override most restrictions,
and the changes are made with an eye towards the refactored scheduler PR.

**Comment by ronen-fr** [comment]
Teuthology tests (95 relevant tests) OK.

File: src/osd/scrubber/osd_scrub.cc [hunk 1]
```diff
@@ -69,10 +69,11 @@ bool OsdScrub::scrub_random_backoff() const
 
 void OsdScrub::initiate_scrub(bool is_recovery_active)
 {
-  if (scrub_random_backoff()) {
-    // dice-roll says we should not scrub now
-    return;
-  }
+  const utime_t scrub_time = ceph_clock_now();
+  dout(10) << fmt::format(
+		  "time now:{}, recover is active?:{}", scrub_time,
```
Comment (athanatos) [inline src/osd/scrubber/osd_scrub.cc]: recovery

File: src/osd/scrubber/osd_scrub.cc [hunk 2]
```diff
@@ -157,39 +137,52 @@ void OsdScrub::initiate_scrub(bool is_recovery_active)
 }
 
 
-std::optional<Scrub::OSDRestrictions> OsdScrub::restrictions_on_scrubbing(
+Scrub::OSDRestrictions OsdScrub::restrictions_on_scrubbing(
     bool is_recovery_active,
     utime_t scrub_clock_now) const
 {
-  // our local OSD may already be running too many scrubs
+  static_assert(sizeof(Scrub::OSDRestrictions) <= sizeof(uint32_t));
```
Comment (athanatos) [inline src/osd/scrubber/osd_scrub.cc:143]: Why is this important?

File: src/osd/scrubber_common.h [hunk 3]
```diff
@@ -49,11 +49,14 @@ enum class scrub_prio_t : bool { low_priority = false, high_priority = true };
 using act_token_t = uint32_t;
 
 /// "environment" preconditions affecting which PGs are eligible for scrubbing
+/// (note: struct size should be kept small, as it is copied around)
```
Comment (athanatos) [inline src/osd/scrubber_common.h:52]: Remove the above static assert or make sure it's within a few lines of the struct and this explanation.

**Review by athanatos — APPROVED** [review]
LGTM other than the nits.

File: src/osd/scrubber/osd_scrub.cc (hunk 1 above)
Comment (ronen-fr) [reply src/osd/scrubber/osd_scrub.cc]: fixed

File: src/osd/scrubber_common.h (hunk 3 above)
Comment (ronen-fr) [reply src/osd/scrubber_common.h:52]: fixed

**Comment by ronen-fr** [comment]
added a text message fix to the osd-recover-scrub test

**Comment by ronen-fr** [comment]
Merging based on https://pulpito.ceph.com/rfriedma-2023-12-20_08:40:50-rados-wip-rf-pcnt3-distro-default-smithi/
This is a full set of Rados tests. 22 failures (16f, 6dead).
All failures (apart from the 3 deads that are reimaging/infra issues) match 'main':
see https://pulpito.ceph.com/yuriw-2023-12-11_16:41:51-rados-wip-yuri3-testing-2023-12-07-0727-distro-default-smithi/

**Pull request merged**